### PR TITLE
[#171549895] first test

### PR DESCRIPTION
**Pull request by balanza**
This is the first instance of integration test performed on IO app API. It's barebone and includes just an unauthorized call to the service. 

I publish this PR as a draft to have a quick code review

### CODE REVIEW

File: integrations/getUserProfile/index.test.ts [hunk 1]
```diff
@@ -0,0 +1,40 @@
...
+describe("getUserProfile", () => {
+  const backendBaseurl = getEnvValue("IO_BACKEND_HOST");
+  const endpoint = `${backendBaseurl}/profile`;
+
+  test("is unauthorized without sessionToken", async () => {
+    const headers = {};
+    const expectedHttpCode = 401;
+    const expectedBody = "Unauthorized";
+
+    const client = () => fetch(endpoint, { headers });
```
Comment (gunzip) [inline integrations/getUserProfile/index.test.ts]: these calls should use the generated types (using io-utils) + the relative client 
in the case of the app backend you can mutuate it from 

https://github.com/pagopa/io-app/blob/master/ts/api/backend.ts

(I agree this should be generated automatically and put in a package : )

File: lib/env.ts [hunk 2]
```diff
@@ -1,4 +1,7 @@
-type EnvKey = 'SPID_SESSION_TOKEN';
+export type EnvKey = (
```
Comment (gunzip) [inline lib/env.ts]: I suggest to have this as a [io-ts](https://github.com/gcanti/io-ts) type so you can deserialize the needed vars using

`EnvKey.decode(process.env)`

File: setup/index.ts [hunk 3]
```diff
@@ -1,11 +1,16 @@
-import { askSessionToken } from './prompt';
-import { get as getEnvValue, set as setEnvValue } from './env';
+import { /* askSessionToken, */ askIOBackendBaseurl } from "./prompt";
+import { get as getEnvValue, set as setEnvValue, EnvKey } from "../lib/env";
+
+const promptIfNot = async (key: EnvKey, fn: Function) => {
+  if (!getEnvValue(key)) {
+    const value = await fn();
```
Comment (gunzip) [inline setup/index.ts]: while it may be in this specific "restricted" case, generally speaking it's better to avoid callbacks when possible, they make things complex and hard to type :)

this kind of pattern(s) is better implemented with an [Option](https://github.com/gcanti/fp-ts/blob/1.12.0/src/Option.ts)
ie. `const myvalue = fromNullable(process.env[<myvar>]).getOrElseL(...)`

File: setup/prompt.ts [hunk 4]
```diff
@@ -15,3 +15,19 @@ export const askSessionToken = async () => {
 
   return sessionToken;
 };
+
+export const askIOBackendBaseurl = async () => {
```
Comment (gunzip) [inline setup/prompt.ts]: there's no need prompt the user, reading from env suffices.

Comment (balanza) [reply setup/prompt.ts]: Ok.
I'd leave it anyway because is works already with env variables, but it might be better DX for someone

File: integrations/getUserProfile/index.test.ts [hunk 5]
```diff
@@ -0,0 +1,40 @@
+import fetch from "node-fetch";
+
+import { get as getEnvValue } from "../../lib/env";
+
+describe("getUserProfile", () => {
```
[the remark this answers is not in the document]
Comment (balanza) [reply integrations/getUserProfile/index.test.ts]: I like to use `operationId`s because they refer to OpenAPI identifiers for each endpoint specification. I consider the actual url of the endpoint an implementation detail, while the actual name is more meaningful for our domain. 
Also, I plan to automate over the spec file in next stages of this suite, and `operationId` would be better reference than actual paths.

if you think it's not adding any issue, I'd leave this.

File: integrations/getUserProfile/index.test.ts (hunk 1 above)
Comment (balanza) [reply integrations/getUserProfile/index.test.ts]: I didn't for a couple of reasons:

1. `io-utils`'s type system enforce me to not call an endpoint in a way it shouldn't (example: cannot call an authenticated endpoint without authentication). This is good at application level, but is not here because it doesn't allow me to stress the service with unexpected calls.
2. I'd prefer to leave the test suite as low-level as I can. In the end, the subject of tests is a restful service, thus I want each test to express clearly what kind of http call is going to perform.

File: setup/index.ts (hunk 3 above)
Comment (balanza) [reply setup/index.ts]: I resolve this conversation as the code is outdated

File: integrations/getUserProfile/index.test.ts (hunk 1 above)
Comment (gunzip) [reply integrations/getUserProfile/index.test.ts]: That's a very good point ! While I agree that it's ok to use your own client (to force invalid requests), in case of successful calls that would be useful to check that the response payloads aligns to the type the client is expecting from the server (and the client is using the generated types). This is useful even to check that there are no API breaking changes deployed. What about a compromise ? (use your own client for calls, check the response payload against generated types for successful calls)

Comment (balanza) [reply integrations/getUserProfile/index.test.ts]: @gunzip yes, that's actually what I meant to do (but I'm having some type-troubles thou 😓).
So:
1. I perform _good ol'_ http request
2. I check `response` object's http code
3. I check if the response body (if any) respects the structure defined by the spec - using generated types and `io-ts` decoders.

How does is sound? Am I missing something?

Comment (gunzip) [reply integrations/getUserProfile/index.test.ts]: sounds right !

File: lib/env.ts (hunk 2 above)
Comment (balanza) [reply lib/env.ts]: Is this the expected solution?

```
export const get = (key: EnvKey): Option<string> =>
  EnvKey.decode(process.env).fold(
    _ => none,
    v => some(v)
  );
```

File: integrations/getApiUserProfile/index.test.ts [hunk 6]
```diff
@@ -0,0 +1,14 @@
...
+  testInvalidToken,
+  testNoToken
+} from "../../lib/helpers/testUnauthorized";
+
+describe("getApiUserProfile", () => {
+  const host = ensureEnvValue("IO_BACKEND_HOST");
+  const basePath = ensureEnvValue("IO_BACKEND_BASEPATH");
+  const endpoint = `${host}${basePath}/api-profile`;
+
+  testNoToken(it, endpoint);
```
Comment (gunzip) [inline integrations/getApiUserProfile/index.test.ts]: do not hide it() calls, make them explicit here

File: integrations/getServices/index.test.ts [hunk 7]
```diff
@@ -0,0 +1,14 @@
+import { ensure as ensureEnvValue } from "../../lib/env";
+import {
+  testInvalidToken,
+  testNoToken
+} from "../../lib/helpers/testUnauthorized";
+
+describe("getServices", () => {
+  const host = ensureEnvValue("IO_BACKEND_HOST");
```
Comment (gunzip) [inline integrations/getServices/index.test.ts]: we have a side-effet here (`throw`) which is implicit. try to never throw or make the fact that you're throwing explicit. it's a complex task in JS to know in advance which methods may throw or not. prefer using eithers / taskeithers.

File: setup/index.ts [hunk 8]
```diff
@@ -1,11 +1,31 @@
...
+  askIOBackendHost,
+  askIOBackendBasePath
+} from "./prompt";
+import { get as getEnvValue, set as setEnvValue, EnvKey } from "../lib/env";
+import { TaskEither } from "fp-ts/lib/TaskEither";
+import { task } from "fp-ts/lib/Task";
+
+const promptIfNot = (
+  key: EnvKey,
+  promptFn: () => TaskEither<Error, string>
```
Comment (gunzip) [inline setup/index.ts]: while it's ok to return monadic structures as output (option, eithers, ...) try to use plain values when possible as arguments (otherwise you're this forcing the caller to "lift" the arguments structures). in this case a plain async method would have been ok (but as we said, try to not pass callbacks :-)

what you're plrobably try to accomplish here can be expresses more explicitely (without hiding the fact that you're changing process.env global object and without modifying it): <--never mutate global objects, try not to mutale local objects:

```ts
const ask = (key: string, message: string, env = process.env) =>
  fromIO(io.of(fromNullable(env[key]))).chain(() =>
    tryCatch(async () => singlePrompt({ message }), toError)
  );

const config = sequenceS(T.taskEither)({
  bar: ask("bar", "prompt message.."),
  foo: ask("foo", "prompt message..")
});

config.map(console.log).run();

// result ->
// { 
//     bar: ' ... value from singlePrompt...',
//     foo: 'value from singlePrompt...' 
// }
```

File: setup/index.ts [hunk 9]
```diff
@@ -1,11 +1,31 @@
...
+) => {
+  const setter = () =>
+    promptFn().fold<string>(
+      error => { throw error },
+      value => {
+        setEnvValue(key, value);
```
Comment (gunzip) [inline setup/index.ts]: this method has a sideeffect (assign to process.env)

File: setup/index.ts [hunk 10]
```diff
@@ -1,11 +1,31 @@
...
+    );
+  return getEnvValue(key).fold(setter(), task.of);
+};
 
 const testSuiteSetup = async () => {
-    if(getEnvValue('SPID_SESSION_TOKEN')) return;
-    
-    const sessionToken = await askSessionToken();
-    setEnvValue('SPID_SESSION_TOKEN', sessionToken);
+  await promptIfNot("IO_BACKEND_HOST", askIOBackendHost).run();
```
Comment (gunzip) [inline setup/index.ts]: try to chain calls and call run() only once at the end. use sequence(T/S) methods and or traverse() to deal with arrays of taskeithers.

Comment (gunzip) [reply setup/index.ts]: see https://github.com/gcanti/fp-ts/issues/1140#issuecomment-589217217

File: package.json [hunk 11]
```diff
@@ -7,17 +7,27 @@
...
-    "test": "jest"
+    "start": "jest",
+    "generate": "npm-run-all generate:api:*",
+    "generate:api:backend": "rimraf generated/backend && mkdir -p generated/backend && gen-api-models --api-spec https://raw.githubusercontent.com/pagopa/io-backend/master/api_backend.yaml --no-strict --out-dir generated/backend --request-types --response-decoders",
+    "generate:api:backend-definitions": "rimraf generated/definitions/backend && mkdir -p generated/definitions/backend && gen-api-models --api-spec https://raw.githubusercontent.com/pagopa/io-backend/master/api_backend.yaml --out-dir ./generated/definitions/backend --no-strict --request-types --response-decoders  && gen-api-models --api-spec https://raw.githubusercontent.com/pagopa/io-backend/master/api_public.yaml --out-dir ./generated/definitions/backend"
   },
   "dependencies": {
     "enquirer": "^2.3.4",
-    "jest": "25.1.0"
+    "fp-ts": "1.12.0",
```
[the remark this answers is not in the document]
Comment (balanza) [reply package.json]: Ok, I just used what found in other packages

File: integrations/getUserProfile/index.test.ts (hunk 1 above)
Comment (balanza) [reply integrations/getUserProfile/index.test.ts]: I resolve the comment as I implemented this is `getServerInfo` suite

File: integrations/getServices/index.test.ts (hunk 7 above)
Comment (balanza) [reply integrations/getServices/index.test.ts]: I can just place the `getOrElseL` here. Is duplication preferable in this case?

Comment (gunzip) [reply integrations/getServices/index.test.ts]: that's ok to put `getOrElseL` here

File: setup/index.ts (hunk 9 above)
Comment (balanza) [reply setup/index.ts]: I originally stored configurations in a Map on a singleton module, with env variables being default values. 
The problem I ran into is jest did not to consider the module a singleton. Maybe because i'm using the setup script as `globalSetup`. So this _workaround_ worked :)

Comment (gunzip) [reply setup/index.ts]: there should be no difference between a singleton module and a non-singleton one since modules shouldn't have side effects

File: setup/prompt.ts [hunk 12]
```diff
@@ -1,17 +1,33 @@
...
     }
-  });
+  }).then((e: Partial<{ value: string }>) => e.value || "");
 
-  return sessionToken;
-};
+export const askSessionToken = () =>
+  tryCatch<Error, string>(
+    () => singlePrompt({ message: "Insert a valid session token" }),
+    (r: unknown) => new Error()
```
Comment (gunzip) [inline setup/prompt.ts]:
```suggestion
    (r: unknown) => new Error()
```
```suggestion
    toError
```

File: lib/env.ts [hunk 13]
```diff
@@ -0,0 +1,65 @@
...
+  Object.keys(obj).reduce<EnvMap>(
+    (p, k: string) =>
+      EnvKey.decode(k).fold(
+        () => p,
+        (key: EnvKey) => ({ ...p, [key]: obj[key] })
+      ),
+    {}
+  );
+
+class Env {
```
Comment (gunzip) [inline lib/env.ts]: > do not use classes unless absolutely necessary; keep data (structures) and behaviors (functions) separate

https://github.com/pagopa/io-handbook/blob/master/development-guidelines.md#general-guidelines

File: lib/env.ts [hunk 14]
```diff
@@ -0,0 +1,65 @@
...
+  t.literal("IO_BACKEND_HOST"),
+  t.literal("IO_BACKEND_BASEPATH")
+]);
+
+export type EnvKey = t.TypeOf<typeof EnvKey>;
+
+export const EnvMap = t.record(EnvKey, t.string);
+export type EnvMap = Partial<t.TypeOf<typeof EnvMap>>;
+
+const pick = (obj: any) =>
```
Comment (gunzip) [inline lib/env.ts]: naming: it looks like a generic method but at the same time it works only with EnvMap / EnvKey. a more specific method and a comment would clarify what this method is supposed to do

File: lib/env.ts [hunk 15]
```diff
@@ -0,0 +1,65 @@
...
+  private constructor(args: EnvMap) {
+    const penv = pick(process.env);
+    this.env = { ...penv, ...args };
+  }
+
+  static instance() {
+    return new Env({});
+  }
+
+  get(key: EnvKey): Option<string> {
```
Comment (gunzip) [inline lib/env.ts]: avoid using getter and setters if not absolutely necessary, they hide what's happening

File: lib/env.ts [hunk 16]
```diff
@@ -0,0 +1,65 @@
...
+    });
+  }
+}
+
+export const get = (key: EnvKey): Option<string> => Env.instance().get(key);
+
+export const set = (key: EnvKey, value: any): void =>
+  Env.instance().set(key, value);
+
+export const setAll = (env: EnvMap): void =>
```
Comment (gunzip) [inline lib/env.ts]: same comment as for "pick"

File: lib/env.ts [hunk 17]
```diff
@@ -0,0 +1,65 @@
...
+  Object.entries(env).forEach(([key, value]) =>
+    EnvKey.decode(key).fold(
+      () => {},
+      k => set(k, value)
+    )
+  );
+
+export const ensure = (key: EnvKey): string | never =>
```
Comment (gunzip) [inline lib/env.ts]: naming: ensure what ? =)

File: setup/index.ts [hunk 18]
```diff
@@ -1,11 +1,31 @@
...
+} from "./prompt";
+import {
+  get as getEnvValue,
+  setAll as setAllEnvValues,
+  EnvKey,
+} from "../lib/env";
+
+const ask = (key: EnvKey, promptFn: () => Promise<string>) => {
+  return getEnvValue(key).fold(promptFn, value => () =>
+    Promise.resolve(value)
```
Comment (gunzip) [inline setup/index.ts]: return TaskEither not promises

File: setup/index.ts (hunk 9 above)
Comment (balanza) [reply setup/index.ts]: Sorry but I don't understand if the problem is that I'm changing `process.env` or it's the assignment itself

Comment (gunzip) [reply setup/index.ts]: first one. I don't get "jest did not to consider the module a singleton" (but we can go on anyway : )

Comment (balanza) [reply setup/index.ts]: ok ok, thus resolved ;)

File: lib/env.ts [hunk 19]
```diff
@@ -0,0 +1,43 @@
+import { Option, fromNullable } from "fp-ts/lib/Option";
+import * as t from "io-ts";
+
+export const EnvKey = t.union([
```
Comment (gunzip) [inline lib/env.ts]: > Use keyof instead of union when defining a union of string literals

https://github.com/gcanti/io-ts#union-of-string-literals

File: lib/env.ts [hunk 20]
```diff
@@ -0,0 +1,43 @@
...
+        (key: EnvKey) => ({ ...p, [key]: obj[key] })
+      ),
+    {}
+  );
+
+const env: EnvMap = { ...pick(process.env) };
+
+export const get = (key: EnvKey): Option<string> => fromNullable(env[key]);
+
+export const set = (key: EnvKey, value: any): void => (env[key] = value);
```
Comment (gunzip) [inline lib/env.ts]: setter methods should return a new object instead of mutating the input

```suggestion
export const set = (key: EnvKey, value: any): void => ({ ...env, [key] = value });
```

never mutate arguments

File: lib/env.ts (hunk 14 above)
Comment (balanza) [reply lib/env.ts]: you're right, it was originally named `pickEnvKeys`, but I crewed that after refactor 😅

Comment (balanza) [reply lib/env.ts]: Anyway, this seems a more general problem, does `io-ts` have any helper to deal with it?

Comment (gunzip) [reply lib/env.ts]: to do what exactly ?

File: lib/env.ts (hunk 16 above)
Comment (balanza) [reply lib/env.ts]: I'd say `get`, `set` and `setAll` are fine in the context they're declared (they refer to `Env` module). Anyway, when I use them I alias such methods

File: setup/index.ts (hunk 18 above)
Comment (balanza) [reply setup/index.ts]: I think I need some more expertise in `fp-ts` to appreciate the full value of this.

Thanks 💪

Comment (gunzip) [reply setup/index.ts]: https://twitter.com/gabro27/status/1085105462096343041

summary: taskeither are lazy and have the same semantics as the other monadic structures (ie. eithers): you can seamless chain them, mapping, folding, etc.. you don't need special constructures (ie. async / await) to work with them since these structures (monads) share the same interface.

there's a very good blog series by gcanti here:
https://dev.to/gcanti/getting-started-with-fp-ts-io-36p6

File: setup/index.ts (hunk 10 above)
Comment (balanza) [reply setup/index.ts]: I'm trying applying this in many ways, but I cannot get it right. Sure I can group each task and run them at once, but I cannot have them to wait for the previous to be completed. This

```
  const backendHostTask = ask("IO_BACKEND_HOST", askIOBackendHost);
  const backendBasepathTask = ask("IO_BACKEND_BASEPATH", askIOBackendBasePath);

  const seq = array.sequence(taskEither);
  return seq([backendHostTask, backendBasepathTask])
    .fold(
      () => {
        throw requiredParamError();
      },
      value => value
    )
    .map(([IO_BACKEND_HOST, IO_BACKEND_BASEPATH]) => {
      setAllEnvValues({
        IO_BACKEND_HOST,
        IO_BACKEND_BASEPATH
      });
    })
    .run();
```

This is basically an implementation of `Promise.all`, where each task is run in parallel. How to have them run one after each other?

Comment (gunzip) [reply setup/index.ts]: https://grossbart.github.io/fp-ts-recipes/#/async?a=run-a-list-of-tasks-in-sequence

Comment (balanza) [reply setup/index.ts]: thanks

File: lib/env.ts (hunk 14 above)
Comment (balanza) [reply lib/env.ts]: Basically what I did with my custom `pickEnvKeys` function.

File: lib/env.ts (hunk 16 above)
Comment (gunzip) [reply lib/env.ts]: why force the user to alias the methods ?

File: lib/env.ts (hunk 20 above)
Comment (balanza) [reply lib/env.ts]: The whole point of using this singleton is having a shared store between the setup script and the actual jest test. I can do what you suggest, yet I have to commit such changes somewhere. Suggestions?

File: lib/env.ts (hunk 14 above)
Comment (gunzip) [reply lib/env.ts]: it's not very clear from the typings (since it takes an "any" and the return type is not specified). but I guess it's supposed to extract some values, identified by a list of literals, from process.env.

What about `EnvMap.decode(process.env).map(config => ...)` <-- you have all the needed values here

Comment (balanza) [reply lib/env.ts]: `EnvMap.decode(process.env)` would fail, as `process.env` has may more keys. Anyway, we can live without that for now

Comment (gunzip) [reply lib/env.ts]: it won't

Comment (gunzip) [reply lib/env.ts]: I'd suggest to drop the whole module and use something like

File: lib/env.ts (hunk 20 above)
Comment (gunzip) [reply lib/env.ts]: just export a constant with all the values or use the solution in https://github.com/pagopa/io-api-testing/pull/1#discussion_r387736861 if you need to get them dynamically
(the part that I didn't get is on why jst would fail if the module is not seen as a singleton)

File: lib/env.ts (hunk 14 above)
Comment (balanza) [reply lib/env.ts]: this fails:

```
const toEnvMap = (obj: any) =>
  EnvMap.decode(obj).getOrElseL(() => {
    throw new Error("decode failed");
  });
const envMap = toEnvMap(process.env); // err: decoded failed
```

Comment (gunzip) [reply lib/env.ts]: that's because it's a record, interface have a loose validation

Comment (balanza) [reply lib/env.ts]: yes, but still I'm getting **all** environment variables if I merge the object. That is: with interface, what I get at runtime is not what I get in the type system

File: lib/env.ts (hunk 20 above)
Comment (balanza) [reply lib/env.ts]: so you just 
```ts
import * as env from './lib/env';
console.log(env.HOME);
env.HOME = 'new home';
```
?

_about the singleton part_
nodejs modules are singleton. Maybe it was just me but now it's fine (data is stored into `env` constants inside `lib/env.ts` module, which is allocated for the entire life of the application)

File: lib/env.ts (hunk 14 above)
Comment (gunzip) [reply lib/env.ts]: why / when do you need to merge the object ? (which object ?)

File: lib/env.ts (hunk 20 above)
Comment (gunzip) [reply lib/env.ts]: why do you need to reassign env ? (generally speaking, if you need to mutate creaet another object)

File: lib/env.ts (hunk 14 above)
Comment (balanza) [reply lib/env.ts]: I'm getting defaults from `process.env` at startup, then I integrate what's missing with values from prompt. 
So I'm doing `const env: EnvMap = {...process.env}`, but that brings a lot of other variables. So the need for `pick` function

Comment (gunzip) [reply lib/env.ts]: Now I get it :-) In this case you can use t.exact:

```ts

import * as t from "io-ts";

const ConfigT = t.exact(t.interface({ HOME: t.string, USERNAME: t.string }));

const config = ConfigT.decode(process.env).getOrElseL(() => {
  throw Error("...");
});

console.log(config);

```

File: lib/env.ts (hunk 20 above)
Comment (balanza) [reply lib/env.ts]: ok, but still this object has to be stored in a _mutable_ location, isn't it?

Comment (gunzip) [reply lib/env.ts]: can you please elaborate it a little bit more ? (what do you intend to be stored into a mutable location)

**Comment by gunzip** [comment]
@balanza I'll try to explain my major concern here (not an esy task :-)

We use a set of common tools, concept and (functional) patterns (map, fold), structures (either, taskeither, option, ...) which are a common knowledge shared between developers.

I notice here a trend to "hide" this shared knowledge behind some utility methods that add a little value. ie.

It's better to use this code "inline" (when needed)

```ts
ConfigT.decode(process.env).getOrElseL(() => {
  throw Error("...");
});
```

than to have some method (an alias)

```ts
const myMethod = () => ConfigT.decode(process.env).getOrElseL(() => {
  throw Error("...");
});
```

which means the DEV must lookup `myMethod` which is expecially hard to get without comments and/or exhaustive typings (ie. it uses "any"). If you multiplicate this effort for a lot of methods, the code becomes complex to understand.

So my suggestion is: try hard to be explicit on your intent :-)

File: lib/env.ts (hunk 20 above)
Comment (balanza) [reply lib/env.ts]: From beginning: I have a store, `lib/env.ts`, in which I have **_roughly_** this:
```ts
const env = {};
export const get = (key) => env[key];
export const set = (key, value) => (env[key] = value)
```
The `env` variable is the _mutable location_ i'm referring to. Even if I change `set` into something like
```ts
export const set = (key, value) => ({ ...env, [key]: value });
```
I will then have to
```ts
import { set } from './lib/env.ts';
const env1 = set('foo', 'bar');
// --> where do I save env1?
```

**Comment by balanza** [comment]
> @balanza I'll try to explain my major concern here (not an esy task :-)
> 
> We use a set of common tools, concept and (functional) patterns (map, fold), structures (either, taskeither, option, ...) which are a common knowledge shared between developers.
> 
> I notice here a trend to "hide" this shared knowledge behind some kind of utility methods that add a little value. ie.
> 
> It's better to use this code "inline" (when needed)
> 
> ```ts
> ConfigT.decode(process.env).getOrElseL(() => {
>   throw Error("...");
> });
> ```
> 
> than to have some method (an alias)
> 
> ```ts
> const myMethod = () => ConfigT.decode(process.env).getOrElseL(() => {
>   throw Error("...");
> });
> ```
> 
> which means the DEV must lookup `myMethod` which is expecially hard to get without comments and/or exhaustive typings (ie. it uses "any"). If you multiplicate this effort for a lot of methods, the code becomes complex to understand.
> 
> So my suggestion is: try hard to be explicit on your intent :-)

File: lib/env.ts (hunk 20 above)
Comment (gunzip) [reply lib/env.ts]: what do you mean with "save" ?

Comment (balanza) [reply lib/env.ts]: I'm creating a new version of the "state", I have to store it and make it available to others

**Comment by balanza** [comment]
This is a very important principle and I think we should pin this comment and make it available also in the [styleguide](https://github.com/pagopa/io-handbook/blob/master/development-styleguide.md).

> @balanza I'll try to explain my major concern here (not an esy task :-)
> 
> We use a set of common tools, concept and (functional) patterns (map, fold), structures (either, taskeither, option, ...) which are a common knowledge shared between developers.
> 
> I notice here a trend to "hide" this shared knowledge behind some kind of utility methods that add a little value. ie.
> 
> It's better to use this code "inline" (when needed)
> 
> ```ts
> ConfigT.decode(process.env).getOrElseL(() => {
>   throw Error("...");
> });
> ```
> 
> than to have some method (an alias)
> 
> ```ts
> const myMethod = () => ConfigT.decode(process.env).getOrElseL(() => {
>   throw Error("...");
> });
> ```
> 
> which means the DEV must lookup `myMethod` which is expecially hard to get without comments and/or exhaustive typings (ie. it uses "any"). If you multiplicate this effort for a lot of methods, the code becomes complex to understand.
> 
> So my suggestion is: try hard to be explicit on your intent :-)

File: lib/env.ts [hunk 21]
```diff
@@ -1,42 +1,42 @@
...
+export const EnvMap = t.exact(
+  t.partial({
+    SPID_SESSION_TOKEN: t.string,
+    IO_BACKEND_HOST: t.string,
+    IO_BACKEND_BASEPATH: t.string
+  })
+);
+export type EnvMap = t.TypeOf<typeof EnvMap>;
 
 const pickEnvKeys = (obj: any) =>
```
Comment (gunzip) [inline lib/env.ts]: can we replace this whole module with a call to `EnvMap.decode(process.env)`  when needed ?

File: setup/index.ts [hunk 22]
```diff
@@ -1,11 +1,54 @@
...
+  getEnvValue,
+  setAllEnvValues,
+  EnvMap
+} from "../lib/env";
+
+import { tryCatch, taskEitherSeq } from "fp-ts/lib/TaskEither";
+import { toError } from "fp-ts/lib/Either";
+import { array } from "fp-ts/lib/Array";
+
+const ask = (key: keyof EnvMap, promptFn: () => Promise<string>) => {
```
Comment (gunzip) [inline setup/index.ts]: avoid callbacks

File: setup/prompt.ts [hunk 23]
```diff
@@ -1,17 +1,23 @@
...
-  return sessionToken;
-};
+export const askSessionToken = () =>
+  singlePrompt({ message: "Insert a valid session token" });
+
+export const askIOBackendHost = () =>
+  singlePrompt({ message: "Insert the IO Backend host" });
+
+export const askIOBackendBasePath = () =>
+  singlePrompt({ message: "Insert the IO Backend base path" });
```
Comment (gunzip) [inline setup/prompt.ts]: there's no need to alias `singlePrompt()`, utility methods must add value, just use it inline when needed

Comment (balanza) [reply setup/prompt.ts]: Aliases avoid prompt details to leak into the `index.ts` ("main") module, which don't necessary need to know details such message, validation or even the source of the input. 
(it's just a coincidence here that all three inputs share the same configuration).

Although this conflicts with the principle of "make the code explicit", it stands for the principle of "keep each code unit at the same level of abstraction".

Comment (gunzip) [reply setup/prompt.ts]: that's exactly the main module that should know about message prompts. exactly where getopts belong.

File: setup/index.ts (hunk 22 above)
Comment (balanza) [reply setup/index.ts]: done

File: lib/env.ts (hunk 21 above)
Comment (balanza) [reply lib/env.ts]: You mean exposing 
```ts
export const env: EnvMap = { ...pickEnvKeys(process.env) };
```
and use it like 
```ts
const host = env.IO_BACKEND_HOST;
env.IO_BACKEND_HOST = askSessionToken();
```
?

Comment (balanza) [reply lib/env.ts]: Is the same think we're discussing [here](https://github.com/pagopa/io-api-testing/pull/1#discussion_r387803853)?

File: setup/prompt.ts (hunk 23 above)
Comment (balanza) [reply setup/prompt.ts]: ok

File: lib/env.ts (hunk 21 above)
Comment (gunzip) [reply lib/env.ts]: i mean `export const config = EnvMap.decode(process.env).getOrElse(...)`

it's not clear why do you need to reassign

```
const host = env.IO_BACKEND_HOST;
env.IO_BACKEND_HOST = askSessionToken();
```

can't you jst use `const sessionToken = askSessionToken()` where's needed ?

Comment (balanza) [reply lib/env.ts]: no, because I need such information in another module of another workflow (the jest test runner)

Comment (gunzip) [reply lib/env.ts]: Now I get it ! You are into an edge case where you need a global setup for jest that takes input from command line (if I remember well one of the initial suggestion was just take the values from environment and drop the cli prompts as tests usually runs inside CI environment). You probably need some jest facility like this: https://jestjs.io/docs/en/configuration.html#globalsetup-string
but it depends on [jest support](https://github.com/facebook/jest/issues/7184). I renew my suggestion to just use environment variables for configuration.

Comment (balanza) [reply lib/env.ts]: `globalSetup` is what I'm using. The difference from the given example is that I built a custom global store instead of accessing the `global` object directly. I did it for implementing the type-driven logic we long discussed (`type EnvMap`).

**About the prompt:**  
The prompt itself is just a toy, I think we'll use it very rarely. What I do care about is the architecture is built upon, which may enable near-future improvements to the test suite like data seeding, data enrichment, automatic token retrieval, data cleanup, service start and teardown, etc.
If the cost is to have an assignment in a module, I advocate we keep it as it might make such feature easier to implement. Yes, this is very speculative and so close to _premature optimization_ pitfall. I leave the choice to you thou. 

If you are worrying about having the service startup complicates, please consider that the jest runner itself is build to only care about env variables. So if we want to get rid of the setup part we can just remove `globalSetup: './setup/index.ts'` from `jest.config.js`, and the test suite keeps working as expected.

File: lib/env.ts (hunk 20 above)
Comment (balanza) [reply lib/env.ts]: the conversation continues here: https://github.com/pagopa/io-api-testing/pull/1#discussion_r388394446

File: lib/env.ts (hunk 21 above)
Comment (gunzip) [reply lib/env.ts]: about 

> I think we'll use it very rarely

the suggestion is [this](https://it.wikipedia.org/wiki/You_aren%27t_gonna_need_it). 
best code is no code. please remove the code about prompts.

Comment (balanza) [reply lib/env.ts]: Ok.

**Comment by balanza** [comment]
I closed the PR as it's duplicated by #4
